### ISSUE TITLE
Conntrack issues

### ISSUE THREAD

**Issue by cozybear-dev** [Enkidu-6/tor-ddos#57]
Hi,

I'm running your iptables (not your latest release from yesterday, but the one before that) and have trouble with the open connections. It was at a critical level (maxed out) at one point, so I flushed everything and reconfigured it all - but it's still increasing too quick. Do you have any suggestions or idea where it may come from? Configuration issue at my side?

These are my conntrack stats over time;
![image](https://user-images.githubusercontent.com/7195866/211157238-588542c9-cea4-4277-bcb6-a24274b2719a.png)

**Comment by cozybear-dev** [comment]
When running your conntrack.sh script, it shows my own relay IPv4 IP is responsible for around 80% of all connections.

**Comment by Enkidu-6** [comment]
Please run conntrack.sh and post the results here. Also type
```
iptables -S -t mangle
```
and post the results please.

**Comment by cozybear-dev** [comment]
![image](https://user-images.githubusercontent.com/7195866/211163043-f78b7a37-df07-4550-82be-6a52cb16926a.png)
![image](https://user-images.githubusercontent.com/7195866/211163055-cb216030-aae1-433b-99f1-426a66f5f028.png)

Note; 23.137.249.8 -> IP of the relay/host itself

**Comment by Enkidu-6** [comment]
What are the rules for port 80? Are you running two instances of Tor? [The records](https://metrics.torproject.org/rs.html#details/A6CD6A3242549DF63E57FF00FF89BD899D2AEF8A) show only one relay running at that IP address and on port 443.

33843 is the number of established IPV4 connections you have. Every incoming connection to you and outgoing connections made by you. They seem high I'm assuming, because you're an exit relay. 

If you weren't an exit relay, your conntrack count would have probably been around 4500 instead of 45000. The rest is for the outgoing connections made by you on behalf of the clients when you're browsing the web for them.

Remember, our filters only apply to the incoming connections. We're not restricting the outgoing connections and what people browse in any way.

**Comment by cozybear-dev** [comment]
Port 80 is my dirport; https://github.com/cozybear-dev/tor-node/blob/main/torrc.template#L138

Have I misunderstood, and should the dirport not be included? Only the orport (443)?

I also run unbound (DNS) on my node, that's also uncommon - but don't think that would create issues like these.

**Comment by Enkidu-6** [comment]
Well, the iptables rules are meant to filter connections to your ORPort. Having them for the Dir Port won't hurt but won't do much either. You my want to remove them.

Now, are you telling me that having the iptables rules have caused an increase in your conntrack count? It should really have the opposite effect. We can try a few things to troubleshoot that if that's the case. What was your conntrack count before applying the rules?

**Comment by cozybear-dev** [comment]
I don't have enough analytical data to say anything before I used the iptable rules. Will do some testing without the dirport and also without any rules at all.

Feel free to close the issue for now. I'll reopen if I notice anything odd whilst testing the different configurations. Thank you for the input.

**Comment by Enkidu-6** [comment]
Download the new update.sh
```
wget -O update.sh https://raw.githubusercontent.com/Enkidu-6/tor-ddos/main/update.sh
```
Remove the port 80 from your ipv4.txt and ipv6.txt files and run the update. That will remove port 80.
Check your conntrack counts after may be 10 minutes. Later you can type:
```
iptables -F -t mangle
ip6tables -F -t mangle
ipset destroy
```
That will remove all the rules completely and then run conntrack.sh after 10 minutes. This way we can compare all three conntrack counts and see the difference.

I'll keep this issue open so we can troubleshoot this.

**Comment by cozybear-dev** [comment]
At 14:15 I applied the first step, at 14:42 I applied the second step.

Before (no change);
![image](https://user-images.githubusercontent.com/7195866/211199275-c73602aa-629d-42f1-a83a-257d71cb88af.png)

After 1 (first step);
![image](https://user-images.githubusercontent.com/7195866/211199294-d6bc5fdc-00cb-4ef9-9090-d23324b9eb9e.png)

After 2 (second step);
![image](https://user-images.githubusercontent.com/7195866/211200215-b567dded-600b-4a46-b14e-efaad4e95be0.png)

As you can see, after step 2 -> there is nearly no change in conntrack anymore.

Overview;

![image](https://user-images.githubusercontent.com/7195866/211200275-e547d3aa-90cd-4446-8ff6-46ef220c3795.png)

**Comment by Enkidu-6** [comment]
Hmm... So the last one is with no iptables rules at all? That's strange. Those -j DROP rules should effectively reduce the size of your conntrack table. Although you have more connections on the second one and the difference is not a lot. Still ...

My best guess is that the hashlimit rules cause stale conntrack entries. If you'd be willing to, I'd like to ask you to try one more thing.

Please comment out lines 31 and 45 of update.sh by putting # to the left of those lines and run it again. It's the lines that have 
```
..... -m hashlimit --hashlimit-name TOR-%s-%s --hashlimit-mode ....
```
in them. This will effectively skip those rules and you'll be running the iptables rules without the hashlimit. It would help me with my theory and might help others too.

Thanks

**Comment by cozybear-dev** [comment]
The last one is without iptable rules at all indeed;

![image](https://user-images.githubusercontent.com/7195866/211201856-3d3e1752-e4ce-4f06-a56d-37cb63c456fe.png)
![image](https://user-images.githubusercontent.com/7195866/211201880-bd621a77-aad8-4a31-8a47-26011cc90293.png)

I'll try your suggestion now.

**Comment by cozybear-dev** [comment]
Results thus far;

![image](https://user-images.githubusercontent.com/7195866/211203724-9fab9c25-783f-4913-8a50-d538656fffbf.png)

Will continue this config for now.

**Comment by Enkidu-6** [comment]
Great, Let it run for a while, may be a day, and please let me know if you see an unreasonable increase in your conntrack counts.. Unfortunately I don't have an exit node to test things further on but hashlimit can have its side effects that I don't like. 

There are ways to be 100% sure if that's the cause but for that I need an exit node with the original rules. A couple of commands and looking at the results can confirm this.

Let me know how it goes please.

**Comment by cozybear-dev** [comment]
To give you an update;

![image](https://user-images.githubusercontent.com/7195866/211777585-620e38d2-d232-4c2c-af6e-6502e81e75df.png)

The fall to 0 is because of an unexpected reboot (don't have persistent iptables at the moment). My server automatically reboots if updates are available, so I'm always on the latest releases... For the time it was able to log - nothing weird was found. Will send an update with more data later.

**Comment by cozybear-dev** [comment]
With y-axis;

![image](https://user-images.githubusercontent.com/7195866/211777826-5a853de5-4fe9-4a2c-9744-dac6658d5e14.png)

**Comment by Enkidu-6** [comment]
Thank you for the update. I have a feeling that the initial high value of the conntrack entries that you experienced didn't have much to do with the hashlimit. The ups and downs have more to do with your system being an exit node than anything else. 

The number of clients and their browsing at any given time of the day will obviously effect the conntrack count and as long as it goes back down you don't have a problem with stale entries..

**Comment by cozybear-dev** [comment]
Update; conntrack has been very stable - had no issues thus far.

**Comment by Enkidu-6** [comment]
Excellent. Thanks for the update. Glad it's working out. I removed the hashlimit rules in v5.1.0 for everyone. I never really liked hashlimit. I guess we can close this issue for now. Feel free to reach out if there are any other issues or questions.

**Comment by cozybear-dev** [comment]
Interesting. Yesterday I observed, what I think, is typical DDOS behaviour. I now run two nodes (increasing to six in upcoming days) and I observed this;

Two different servers, sudden sharp increase in conntrack count. The server that is maxed out, has no bandwidth restrictions - the server that does, didn't hit the ceiling. Have you seen this as well? This must be DDOS related right?

![image](https://user-images.githubusercontent.com/7195866/215261758-c513efec-0046-4aaf-b296-fe182a33be98.png)
![image](https://user-images.githubusercontent.com/7195866/215261774-ab697116-4605-4913-94f7-00bbdf880372.png)

**Comment by ToterEngel** [comment]
I can't see any concrete increase in the last few days.
However, my relays have been getting worse and worse over the past few weeks. I even switched off my second relay again.
Even with one relay per server, the load sometimes increases to 25 when an attack starts again.
That's hardly bearable.

**Comment by cozybear-dev** [comment]
![image](https://user-images.githubusercontent.com/7195866/215262115-0d5658e1-4173-4378-bd60-df03fe32b26b.png)
![image](https://user-images.githubusercontent.com/7195866/215262126-30ca7a30-5078-4741-bc34-621ec42b5e00.png)

One of my  nodes was down for a few minutes (the one that hit the ceiling). You can see it in my monitoring tooling as well, timeout increased drastically.

**Comment by Enkidu-6** [comment]
@cozybear-dev  I'm afraid I'm a bit confused by the charts as they have no reference or scale to show me the number of conntrack entries. I can work better with actual numbers and data. Would you please run conntrack.sh and post it please?

@ToterEngel Could you be a bit more specific as what you mean by worst? Is it a spike in conntrack numbers, RAM, CPU? 

My relays have been running very smoothly especially in the past few days. One of them has a bit higher RAM usage, the other one is happily running at a low RAM. Here's a sample of my conntrack.sh:

```
Conntrack count:
   8505
IPs with more than 2 connections:
      3 141.212.118.18
      3 193.187.88.42
      3 193.187.88.43
    699 2001:41d0:0403:4251:0000:0000:0000:0009
   2524 46.105.154.14
Two connections, Relays:
 There are 165 IPs with two connections 162 of which are Relays
Two connections, dual-or:
 There are 165 IPs with two connections 162 of which are Dual-OR Relays
One connection, Relays:
 There are 4221 IPs with one connection 2073 of which are Relays
One connection, dual-or:
 There are 4221 IPs with one connection 268 of which are Dual-OR Relays
```
and if you look at the total number of connected IPs an double it, they roughly match the conntrack count.

**Comment by Enkidu-6** [comment]
@cozybear-dev 
P.S.
What would really help me try to figure this out is a file containing your netstat -aen and a file containing your conntrack entries. I can try to filter them and figure out which connections are responsible for your high conntrack entries. Your issue is different from most of us because you're running an exit relay.

The way conntrack works is that it keeps a flow state for every request. But the interesting part is that any time it sees a -j DROP anywhere, no flow state is created. This is why we selectively have a few -j DROPs in our rules to reduce the bogus conntrack entries. 

The problem is that we're not and we shouldn't filter outgoing packets for an exit relay and the outgoing traffic that you are generating on behalf of the clients, are the main cause of the high conntrack count.

So if or whenever you feel like it create two files by typing:
```
netstat -aen > netstat
cat /proc/net/nf_conntrack > conntrack   # if you have that file 
# Or f you don't:
conntrack -L > conntrack
conntrack -L -f ipv6 >> conntrack
```
and get me those two files so I can play with them and see what's going on on your exit relay that causes this.

**Comment by cozybear-dev** [comment]
@Enkidu-6 

One of my nodes is actually down right now. I'll post the data of my other node once it's up again.

![image](https://user-images.githubusercontent.com/7195866/215267018-0a4646ca-9572-4af3-b4cc-1551922bbe05.png)

![image](https://user-images.githubusercontent.com/7195866/215268071-a8fbb842-73bb-4844-bc14-e8fbfe5cf380.png)

Regarding the request for netdata data, the relevant data you seek is gone by now is it not? I'm not going to share that detailed information if it's not directly usable.

**Comment by Enkidu-6** [comment]
That's understandable. Right now it looks good and reasonable. I do see though that you're most likely not running the newest version. The hashlimit rule that was causing you trouble is long gone in the newer versions. The new rules manage the connections better and instead of allowing everyone 2 connections, they allow only dual-or relays two connections and everyone else will get only one. It also manages SYN packets better which in turn will reduce the number of conntrack entries. You might want to consider giving it a try.

**Comment by ToterEngel** [comment]
When running my relay without limitation it consumes around 8-10 GB of RAM after a few days of operation.
However, this is not a problem, I usually have >50GB of free RAM capacity.

Here is also an overview of the Conntrack statistics for one week.
![fw_conntrack-week](https://user-images.githubusercontent.com/113014221/215269984-e9daa580-b5db-4cb5-9c30-efe5d71cf4af.png)

This is always very even.

The only thing that causes extreme peaks every day is the CPU & system load.
What I can hardly explain myself with an Intel i7, where hardly anything else is running on it.

**Comment by Enkidu-6** [comment]
The CPU spikes you see are caused by these:

![Untitled4](https://user-images.githubusercontent.com/101878139/215270551-9cd8a2be-52e1-42b8-ba10-873f6909235a.jpg)

My Max Advertised Bandwidth is 20 M but as you can see the netin at one time reached 35 M. That's when a bunch of them at the same time flood your system with requests and your CPU and RAM try to deal with them. They are Dropped in our case and nothing goes out but even dropping them requires CPU and RAM usage.

**Comment by cozybear-dev** [comment]
Still not sure what it is precisely. I now host 6 tor exit nodes, all with your latest rulesets. Yesterday some sort of attack was performed. Across all of my nodes at around the same time, the conntrack went through the roof. Some nodes go down for a few seconds because of the kernel dropping packets (as table is full). A quick fix, is just to flush the table - but that's a symptomatic bandaid. 

Here is the cumulative conntrack count of all my exit nodes;

![image](https://user-images.githubusercontent.com/7195866/216826103-537ca0fd-21cd-4e9c-801b-19aacb70f7e0.png)

**Comment by cozybear-dev** [comment]
One of the servers that is maxing out as we speak;

![image](https://user-images.githubusercontent.com/7195866/216826455-93d44c34-c48b-4369-bc86-ecd3086b3c07.png)

**Comment by cozybear-dev** [comment]
Isn't it an option to just increase my maximum conntrack count? I have enough ram to safely double it. Still symptomatic bandaiding though.

**Comment by Enkidu-6** [comment]
Well, technically you're not being attacked, your attacking someone else. Those are most likely the connections you're making on behalf of someone else. I have a feeling that they are using Tor Network to attack someone.

Flushing the table basically means dropping all packets, not pretty. Increasing the nf_conntrack_max and nf_conntrack expect_hashsize would be the proper way of doing it.

And yes, depending on your kernel version the conntrack_max can be set to your RAM in bytes divided by 16384 if your kernel is above 5 or RAM in bytes divided by 16384r divided by 2 if your kernel is 4

And your hashsize is conntrack_max divided by 4. So if you have 8 GB of RAM and kernel 4 then:

8x1024x1024x1024 /16384/2 = 262144 
262144 / 4 = 655356

Based on your RAM you can calculate the proper value. Not sure what operating system you have. You can just search for your OS and how to change the value.

**Comment by Enkidu-6** [comment]
Type:
```
sysctl net.netfilter.nf_conntrack_max
```
and 
```
cat /sys/module/nf_conntrack/parameters/hashsize
```
to see if they gives you a number. If they do, try

```
sysctl -w net.netfilter.nf_conntrack_max=123456
```
replace 123456 with the actual number you want to set. This is not persistent though
Make it persistent is different from system to system. Check to see if you have /etc/sysctl.d/99-sysctl.conf if you do then type:
```
echo "net.netfilter.nf_conntrack_max = 123456" >> /etc/sysctl.d/98-conntrack.conf
```
The file name doesn't matter as long as it starts with a lower number than 99 and ends with .conf

and then 
```
echo "options nf_conntrack expect_hashsize=456789 hashsize=456789" >/etc/modprobe.conf
```
Replace 456789 with the actual hashsize

Reboot.

After reboot try sysctl net.netfilter.nf_conntrack_max and  cat /sys/module/nf_conntrack/parameters/hashsize again to see if the new values took.

No guarantees, those methods work for me. Your OS might be different.

**Comment by cozybear-dev** [comment]
Ok, testing two things right now.

4 of my nodes are having downtime due to conntrack table issues. 

1 of my nodes is now configured with custom kernel params (including your suggestion), and increased ulimit for file descriptor issues with tor -> only reloaded tor

1 of my nodes is now configured with custom kernel params (including your suggestion), and increased ulimit for file descriptor issues with tor -> restarted  tor`fully

The 2 other nodes will struggle for now -> but that's nessecary to see what will work.

Let's see if CPU/RAM can keep up, and if there will be any odd issues...

**Comment by Enkidu-6** [comment]
Sounds good. Let's see how it goes.

By the way when you say "reloaded Tor" do you mean just a -HUP and left the existing conntrack entries intact?

**Comment by cozybear-dev** [comment]
@Enkidu-6 correct

**Comment by cozybear-dev** [comment]
To clarify, afaik a process starts with a certain file desceiptor limit. As I updated said limit, I also wanted to do a node with a full restart.

**Comment by Enkidu-6** [comment]
I wasn't aware that you had file descriptor problems. I thought it was only a problem with your conntrack table filling up. Nevertheless, I'm looking forward to seeing the result.

**Comment by cozybear-dev** [comment]
It's both. The conntrack table fills up + some of my nodes were very near (95%-98%) of total allowed file descriptors. Not always though, sometimes the conntrack table is full before even coming close to the file descriptor limits.

**Comment by cozybear-dev** [comment]
So far so good (above to below; changes + reload, changes + restart, nothing)

![image](https://user-images.githubusercontent.com/7195866/217045543-6e146a0d-02ba-4fbc-9b4b-90e51db1ed7f.png)

**Comment by cozybear-dev** [comment]
Conntrack has been fine for a while now. I now do have some odd issues at 2 of my servers (one of which is high bandwidth - at least 10MiB/s up/down continued), there seems to be some odd configuration issue. Not sure if it's the iptable rules. The behaviour is that my CPU + bandwidth suddenly drop to near zero on a frequent basis. Overall the relay performs fine, but it can't obtain Guard status because of this.

This is what it looks like in stats.

This behaviour is observed on only 2 of my servers (total 10), both on at different hosters and widely different hardware. Any suggestions on what it might me? When it happens often, it's with a minute interval. So perhaps there is some cache config, or other 60s config that is causing this.

![image](https://user-images.githubusercontent.com/7195866/219870481-7873e28f-3b18-4381-a9ee-141636a2e195.png)

**Comment by cozybear-dev** [comment]
I'm currently testing a more clean configuration. I rebooted one of the servers, so all iptables and sysctl config has been wiped. The primary difference in my fleet is that these servers do not have IPv6 addresses, and use an external DNS server - whilst the other servers use unbound.

**Comment by cozybear-dev** [comment]
Since the reboot (running for an hour or something now) the issues have not been seen since.

**Comment by Enkidu-6** [comment]
What do your system logs say? If the CPU and bandwidth drop to zero, there should be some indication in the system logs somehow.

**Comment by cozybear-dev** [comment]
Found nothing odd in the journalctl at least, only continued SSH bruteforcing - which occurs at all my relays and is to be expected. It's as some cache or queue gets full or something like that. Or some delay/limit is being hit.

**Comment by cozybear-dev** [comment]
Perhaps it's the hoster doing some throttling or something.

**Comment by Enkidu-6** [comment]
Some hosts do have some limitations on how many seconds the CPU can consistently run at 100%. Some even have lower limits than 100%. Perhaps you could ask your host if they have any?

**Comment by cozybear-dev** [comment]
Just to give you an update; the servers have been running fine without all the tweaking applied. For now I have the following rule; low bandwidth (lower than 7-8 MBit advertised) apply the tor-ddos ruleset -> all above don't for now. Especially with high bandwidth, I notice the additional performance -> that would otherwise be occupied by overhead of tracking all the connections. Besides that the process jitter is also gone, which was my primary issue.
